wordpress-mobile/WordPress-iOS
# Issue 19051 - Suggestions list disappear after expanding/collapsing the reply input

Closes https://github.com/wordpress-mobile/WordPress-iOS/issues/19051

## Description

In this PR, I'm addressing the requested change to keep the suggestions list visible when expanding/collapsing the reply text field. If the user inserts a `@` the suggestions appear. And if the user clicks on the `^` button, it should be visible when the text area is expanded. Following the same idea, if the text area is expanded and the suggestions list is visible, it should remain visible when the user clicks on the `⌄` button.

## Details

- I created `FullScreenCommentReplyViewModel` to extract business logic from `FullScreenCommentReplyViewController` and make it easy to test the visibility of the suggestions list. 

- I added a new parameter on `FullScreenCommentReplyViewController.enableSuggestions` representing the last search string provided by the user, and it is being called before `FullScreenCommentReplyViewController` is presented. 

- I changed `FullScreenCommentReplyViewController` to display the suggestions list if the search text was provided. The suggestions list is shown or not in `viewDidAppear` using the search text provided. The suggestions list remains hidden if the search text is nil or empty.

- I modified `FullScreenCommentReplyViewController.onExitFullscreen` to have one more string as the last parameter. It represents the last search text inserted by the user when the text input area is expanded. 

- When the user exits `FullScreenCommentReplyViewController`, `onExitFullscreen` completion block is called with the last search text provided. If the user press `Reply` there is no need to keep the list visible, so I'm sending nil in the completion block. 

- I created a new method in `ReplyTextViewDelegate` which is being called when `FullScreenCommentReplyViewController` is closed. The implementation receives the last search text provided by the user, and displays the suggestions list. 

- I'm calling the new delegate method on every view controller that calls `FullScreenCommentReplyViewController` enabling the suggestions list visibility. So, every view controller that conforms with `willEnterFullScreen` method is conforming with the new `didExitFullScreen` method.

## Preview

<p float="left">
  <img src="https://user-images.githubusercontent.com/8026597/186800803-2c37b668-0a44-46eb-8de0-04447a5782a1.gif" width="300" />
  <img src="https://user-images.githubusercontent.com/8026597/186803096-8b9b8092-0c7c-48a2-93a3-b7a5a7ebba94.gif" width="300" /> 
</p>

## Testing Instructions

**N.B: The user-mentions suggestions feature is not enabled in self-hosted sites ( wordpress.org ).**

**Scenario 1 - Suggestions list remains visible when the text area is expanded** 

1. Tap the "My Site" tab.
2. Tap on "Menu".
3. Tap on "Comments" row.
4. Tap any comment.
5. Tap "Reply" button.
6. Type "@" and you should see the user-mentions suggestions list.
7. Click on the `^` button.
8. Check if the suggestions list remains visible. 

**Scenario 2 - Suggestions list remains visible when the text area is collapsed** 

1. Tap the "My Site" tab.
2. Tap on "Menu".
3. Tap on "Comments" row.
4. Tap any comment.
5. Tap "Reply" button.
6. Type "@" and you should see the user-mentions suggestions list.
7. Click on the `^` button.
9. Erase the text to make the suggestions list disappear. 
10. Make it appear again by typing "@".
11. Click on the `⌄` button to collapse the text area.
12. Check if the suggestions list remains visible.

There are other scenarios to display the suggestions list described in https://github.com/wordpress-mobile/WordPress-iOS/pull/18979.

## Regression Notes
1. Potential unintended areas of impact
Unfortunately I couldn't create unit tests for every view controller conforming with the new `didExitFullScreen` delegate method: `CommentDetailViewController`, `NotificationDetailsViewController` and `ReaderCommentsViewController`.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
I've manually tested all the scenarios where `didExitFullScreen` is called. 

3. What automated tests I added (or what prevented me from doing so)
I added `FullScreenCommentReplyViewModelTests` and I added new tests in `FullScreenCommentReplyViewControllerTests` to check the visibility of the suggestions list.
I couldn't add new tests for `CommentDetailViewController`, `NotificationDetailsViewController` and `ReaderCommentsViewController` to ensure the suggestions list is displayed because those view controllers are not unit-testable.

PR submission checklist:
